DNS polling strategy; RFC.

doughsay opened the pull request:
This is an attempt at https://github.com/bitwalker/libcluster/issues/20 that works for my use-case, but makes several assumptions and relies on a polling architecture.  I'd love some feedback on it to see if it could be included in libcluster, or if maybe I should just keep it separate for my own use-case.

Assumptions:
* We poll every once in a while to see if any new nodes with the given dns query have come online, or if any existing ones left
* it assumes you want to do a regular A record query
* it assumes your node names are of the format "node_sname@ip_address"

Like I said, this works for my use-case, but may not really be general enough for inclusion... let me know what you think.

michaelklishin commented:
I can think of a couple of projects that have been using a very similar approach (not with libcluster, by the way) for a few years. Adding a couple of things:

 * AAAA record support
 * Ability to provide a custom address to hostname mapping function

should make this a pretty generic solution.

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+defmodule Cluster.Strategy.DNSPoll do
+  @moduledoc """
```
Could you flesh out the module doc to explain how it works and what options can be used to configure it and any assumptions made?

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+      connect: Keyword.fetch!(opts, :connect),
+      disconnect: Keyword.fetch!(opts, :disconnect),
+      list_nodes: Keyword.fetch!(opts, :list_nodes),
+      config: Keyword.fetch!(opts, :config)
```
I would make this `Keyword.get(opts, :config, [])`, and then down below when you fetch specific options, use `Keyword.fetch!/2` on the require config options, since this will potentially allow someone to pass an empty list for `:config` and get all the way through `init/1` before crashing at runtime because the configuration is actually invalid.

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+    node_sname = Keyword.get(state.config, :node_sname)
+
+    state = %{state | :meta => {poll_interval, query, node_sname}}
+    {:ok, state, 0}
```
I'm not sure if you've encountered this yet, but if a process receives a message during `init` and you specify a timeout like this, the timeout will never occur because a message being received cancels it. It is generally extremely unlikely for that to happen, but I have seen it happen in my own projects, so I've taken to using `Process.send_after/3` any time where that first timeout must occur. One thing I really like about GenStateMachine is that you can insert "internal-only" events to be processed before everything else, it's too bad we don't have that with the other Gen* behaviours!

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+  end
+
+  # timeout starts the loop by calling :poll
+  def handle_info(:timeout, state), do: handle_info(:poll, state)
```
Using `Process.send_after/3` also allows you to get rid of this extra clause by just sending the `:poll` message.

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+  def handle_info(:poll, %State{meta: {poll_interval, query, node_sname}} = state) do
+    debug state.topology, "polling dns for #{query}"
+
+    self = node()
```
Never rebind `self` - it is a built-in function which returns the pid of the current process, and rebinding it will result in bugs.

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+    |> handle_poll(state)
+
+    # reschedule a call to itself in poll_interval ms
+    Process.send_after(self(), :poll, poll_interval)
```
Like this ;)

bitwalker commented on lib/strategy/dns_poll.ex:
```diff
@@ -0,0 +1,75 @@
+  end
+
+  # turn an ip into a node name atom, assuming that all other node names looks similar to our own name
+  defp format_node({a, b, c, d}, sname), do: "#{sname}@#{a}.#{b}.#{c}.#{d}" |> String.to_atom
```
You can just do `:"#{sname}@#{a}.#{b}.#{c}.#{d}` rather than piping to `String.to_atom/1`

palm86 commented:
What happens to global processes in the short period (`< poll_interval`) between a node coming up and it being connected to the others? Say a global process is already running on Node A which is connected to Node B. Now Node C comes up and starts the global process itself, then gets connected to A and B. I suppose it stops the global process on C?

Could this also lead to problems in Phoenix instances running on nodes, with only some of them being connected? Especially thinking about channels.

I guess each BEAM instance could wait for at least poll_interval before doing any significant distributed operations to ensure that it is connected to the cluster first.

bitwalker commented:
> What happens to global processes in the short period (< poll_interval) between a node coming up and it being connected to the others? Say a global process is already running on Node A which is connected to Node B. Now Node C comes up and starts the global process itself, then gets connected to A and B. I suppose it stops the global process on C?

Swarm resolves these conflicts by assigning shards of the keyspace to specific nodes, if a process is assigned to a different node, or its assignment changes, it goes through a handoff process and is shutdown on any nodes it is duplicated on. If a process is unresponsive to the handoff, it is killed by the tracker.

> Could this also lead to problems in Phoenix instances running on nodes, with only some of them being connected? Especially thinking about channels.

For Phoenix, you would almost certainly want to use Phoenix.Presence vs Swarm for channels, a more specific example of the problem you have in mind will help me answer the question though.

> I guess each BEAM instance could wait for at least poll_interval before doing any significant distributed operations to ensure that it is connected to the cluster first.

This is the purpose of the `kernel` application settings, `sync_nodes_mandatory` and `sync_nodes_timeout` - to ensure that all nodes in a static cluster are connected before any applications are started.

palm86 commented:
Thanks for the reply and for mentioning Swarm. It looks very useful! I had no specific problem in mind regarding Phoenix and Channels, I'm just fishing for understanding :)

> This is the purpose of the kernel application settings, sync_nodes_mandatory and sync_nodes_timeout - to ensure that all nodes in a static cluster are connected before any applications are started.

Aren't these settings only useful in a non-autoscaling cluster?

One last question, if you don't mind: when you speak of a network split, do you mean a multiple cluster situation has come about, or do you mean that processes that were supposed to be global/singleton somehow started to run on more than one node in a single cluster?

bitwalker commented:
> Thanks for the reply and for mentioning Swarm. It looks very useful! I had no specific problem in mind regarding Phoenix and Channels, I'm just fishing for understanding :)

I just realized this is my `libcluster` repo, not `swarm` haha, my bad. Coincidentally though, the way Swarm solves it applies to any other library which needs to handle distributed data and potential conflicts, so I guess my comments still work ;)

> Aren't these settings only useful in a non-autoscaling cluster?

Yes and no. You can still use those settings to ensure that the node connects to other members in the cluster, the trick is generating the configuration before boot so that the settings can be used. That said, I'm generally in favor of finding a solution that works with both static and dynamic membership, and `libcluster` was my attempt at finding that. One answer here is that your application could use start phases to start part of your application and then only start the remainder once `libcluster` has established connections to the rest of the cluster and your criteria for what's safe has been met, whether that's a quorum of some number of nodes, or a time period, or something else. You can block indefinitely in a start phase, so it's a safe place to do that kind of waiting. This approach of phased startup is something I recommend, whether you use start phases to do it (only useful for initial boot), or post-initialization within your supervisor tree (useful for anytime the supervisor tree starts or restarts), or some combination of both. This allows your application to fail back to more stable levels of functionality in the case of failure, and also allows you to do things like wait for the environment around you to stablize (whether that's waiting for a cluster, or waiting for other resources). This is especially important if you need to do things like leader election and so on, where you can't just cross your fingers and hope that everything works out when your application boots.

> One last question, if you don't mind: when you speak of a network split, do you mean a multiple cluster situation has come about, or do you mean that processes that were supposed to be global/singleton somehow started to run on more than one node in a single cluster?

A network split implies that some partitioning of the cluster has occurred (but whether the other partitions are dead or alive is unknown, and generally doesn't matter, though your specific use case may need to assume other partitions are alive and take appropriate action based on that knowledge). A network split may also imply that processes which are supposed to be globally unique may now be running in all partitions depending on the behaviour of the application (in the case of Swarm, if a process is assigned to a node and goes down, it is restarted on a new node which has been assigned the old keyspace for the down node, which means all partitions will have a copy of that registered process), and that once the split is healed, you will likely have conflicts which need to be resolved, depending on what that process does. Swarm provides a mechanism for this, which is assigning nodes to part of the keyspace, and when nodes come and go (or a split is healed), we can deterministically say which node is the new "master" for a given process, and then orchestrate the conflict resolution via a handoff mechanism. You can also specify that the now defunct duplicates should just die, e.g. no handoff is required. The downside with Swarm is that it is eventually consistent, which makes it a non-option for some use cases.

There is no "one true way" to deal with these kind of things. It all boils down to the guarantees you need to provide - if you can't allow duplicates during a network split, or you can't afford dirty reads/writes, then you need something strongly consistent with leader/quorum-based decision making, if duplicates are fine, or dirty reads/writes are enough then an eventually consistent system which either resolves conflicts automatically via CRDTs, or manually by orchestrating the conflict resolution proccess may be better as they generally have better performance. So getting back to `libcluster`; the best we can do is provide a way to block startup until a cluster is fully formed (or at least well-formed), but the criteria for that is something dependent on the given application, so I haven't baked anything like that into `libcluster`, instead preferring to point people towards the start phase/post-init approach, and focus solely on the mechanics of forming clusters. Libraries like `Swarm` can make those higher-level decisions since they know more about the use case of the application, but even Swarm avoids doing this (though it was designed for use in a dynamic environment) because you may have other components with different clustering needs.

Sorry, that got overly long, I think that probably was a lot more than you were actually asking, but hopefully the background helps :)

palm86 commented:
Thanks for the awesome reply!

doughsay commented:
@bitwalker Hey, I'm sorry this was lost so long ago.  I think I addressed all the feedback and refactored it a bit.  This has been working for me in production for a while, so I think it can be useful to others.

Note about the `self` being shadowed: for some reason this never caused an issue for me, but I have no idea why...  but this is now fixed as well.

bitwalker commented:
Thanks for the PR! I've got a few tweaks I want to make, but I'll do that in master :)
